rails/database-migrations.md

```diff
@@ -0,0 +1,138 @@
+- Deployment pipelines are usually set up to run data migrations, so there are no extra steps to run.
+
+Of course, there are exceptions to this rule. Super complex data migrations may need more involved scripts.
+
+### Keep migrations runnable for as long as possible
+Migrations that break after only a few git commits are a nightmare for teams. This problem is almost always caused by migration code relying on application code like model associations and methods. A migration runs at a single point in time, whereas application code is always changing, so the two get out of sync.
+
+For example, suppose a migration is added to the master branch which uses an association defined on an ActiveRecord model. Two commits later, the association is removed from the ActiveRecord model, but the migration code isn't updated.
+
+Now, another developer rebases his feature branch onto master and attempts to run the pending migration.
+
+Failure. To be specific, a `NoMethodError: undefined method` failure because the association the migration depended on isn't defined on the model anymore, so the migration blows up. 😩
+
+You can avoid this problem by following a simple rule:
+
+> Migrations may only use code defined within their own class.
+
+If you're migrating data and need helper methods or ActiveRecord associations, define them directly in the migration class itself. See the **Use Cases** section below for an example of how to properly use ActiveRecord associations in a migration.
+
+
+### Do not use migrations to create new data
+Use `seeds.rb` or a rake task to populate necessary app data. Otherwise running `db:schema:load` will leave the app's database without the needed data.
+
+### Always test your migrations
+Always run the migration, roll it back, and re-run it (`rails db:migrate:redo` is handy for this). This prevents accidentally writing irreversible migrations because you forgot to specify all the needed details the `change` method needs to calculate the reverse of the operation you're running.
+
+If a migration by its nature cannot be rolled back, be explicit about it by raising `IrreversibleMigration` in the `down` method.
+
+Other common gotchas to test for:
+- The migration won't run on a fresh database because there was something in the local dev db it was relying on.
+- Use realistic production sized data to check for speed.
+
+## Use Cases
+
+### Use ActiveRecord associations in a migration
+
+Existing model:
+
+```ruby
+class Tag
+  belongs_to :blog_post
+end
+
+class BlogPost
+  has_many :tags
+end
+```
+
+You decide for performance reasons to move the tags into an array column on the `blog_posts` table instead of having them in a separate `tags` table.
+
+Write the migration this way:
+
+```ruby
+class MoveTags < ActiveRecord::Migration[5.2]
+  # Inherit from ActiveRecord::Base instead of ApplicationRecord since the
+  # ApplicationRecord is subject to changes as the app changes.
```

Comment:
Assume that doesn't throw a warning

Reply:
Hmm, that's a good point.

Rails doesn't care or throw a warning, since `ApplicationRecord` is just a simple abstract class:

```ruby
class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true
end
```

Rubocop does have a rule about it though, so Rubocop will complain, although it's easy to disable the check for the `db/migrate` folder. https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Rails/ApplicationRecord

I'm inclined to keep my original recommendation of inheriting from `ActiveRecord::Base` because that's the only way to isolate the migration completely from the constantly changing application code.

Reply:
Could you put an inline Rubocop exclusion in your example migration code?